File: Problems/No.54_Spiral_Matrix/spiral_matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        # parameter
        spiral_order = []

        # start computing
        while matrix:

            # Top Row
            spiral_order += matrix.pop(0)
            
            # Right Column
            if matrix and matrix[0]:
                for row in matrix:
                    spiral_order.append(row.pop())
            
            # Bottom Row
            if matrix:
                spiral_order += matrix.pop()[::-1]
            
            # Left Column
            if matrix and matrix[0]:
                for row in matrix[::-1]:
                    spiral_order.append(row.pop(0))

        return spiral_order

File: Problems/No.54_Spiral_Matrix/spiral_matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        // parameters
        std::vector<int> spiral_order;
        int c_start = 0, r_start = 0;
        int c_end = matrix[0].size()-1;
        int r_end = matrix.size()-1;
        int Msize = (c_end+1)*(r_end+1);

        // start computing
        while(spiral_order.size() != Msize) {

            // Top Row
            for (int i=c_start; i<=c_end; i++) spiral_order.push_back(matrix[r_start][i]);

            // Right Column
            for (int i=r_start+1; i<=r_end; i++) spiral_order.push_back(matrix[i][c_end]);

            // Bottom Row
            if (r_start != r_end) 
                for (int i=c_end-1; i>=c_start; i--) 
                    spiral_order.push_back(matrix[r_end][i]);

            // Left Column
            if (c_start != c_end)
                for (int i=r_end-1; i>=r_start+1; i--) 
                    spiral_order.push_back(matrix[i][c_start]);

            // update parameters
            r_start++;
            c_start++;
            r_end--;
            c_end--;
            
        }

        return spiral_order;

    }
};
